packages/validation/src/validators/metadata/validator.ts

```diff
@@ -0,0 +1,146 @@
+    super(
+      {
+        id: 'MTD',
+        name: 'metadata',
+        displayName: 'Metadata',
+        description: 'Validates that your manifests have correct metadata fields and values.',
+        learnMoreUrl: 'https://kubeshop.github.io/monokle/resource-validation/',
+      },
+      METADATA_RULES
+    );
+  }
+
+  protected preconfigureRules(rules: RuleMap = {}) {
+    // Config method may be called multiple times for same Validator instance
+    // so we need to remove previously added dynamic rules.
+    this._rules = this._rules.filter(rule => !(rule.name.endsWith('-label') || rule.name.endsWith('-annotation')));
+
+    Object.entries(rules).forEach(rule => {
+      const ruleName = rule[0];
+
+      if (!ruleName.startsWith('metadata/') || !ruleName.endsWith('-label') && !ruleName.endsWith('-annotation')) {
+        return;
+      }
+
+      const isLabelRule = this.isLabelRule(ruleName);
+      const ruleTypeName = isLabelRule ? 'label' : 'annotation';
+      const ruleShortName = ruleName.replace('metadata/', '');
+      const ruleNormalizedName = ruleShortName.replace('__', '/').replace((isLabelRule ? /-label$/ : /-annotation$/), '');
+
+      this._rules.push({
+        id: `MTD-${ruleShortName}`,
+        name: ruleShortName,
+        shortDescription: { text: `The ${ruleNormalizedName} ${ruleTypeName} is missing.` },
+        fullDescription: {
+          text: `The resource is violating the ${ruleShortName}. The resource may not be discoverable by tools that rely on these ${ruleTypeName}s.`
+        },
+        help: {
+          text: `Add required ${ruleTypeName}. You can hover the key for documentation.`
+        },
+        defaultConfiguration: {
+          parameters: {name: ruleNormalizedName}
+        },
+      });
+    });
+
+    this.setRules(this._rules);
+  }
+
+  async doValidate(resources: Resource[], incremental?: Incremental): Promise<ValidationResult[]> {
+    const results: ValidationResult[] = [];
+    const dirtyResources = incremental ? resources.filter(r => incremental.resourceIds.includes(r.id)) : resources;
+
+    for (const resource of dirtyResources) {
+      for (const rule of this.rules) {
+        const resourceErrors = this.validateResource(resource, rule);
+        results.push(...resourceErrors);
+      }
+    }
+
+    return results;
+  }
+
+  private isLabelRule(ruleName: string) {
+    return ruleName.endsWith('-labels') || ruleName.endsWith('-label');
+  }
+
+  // We have special behavior for different rules (https://github.com/kubeshop/monokle-saas/issues/722):
+  // MTD-recommended-labels does not allow config parameters (meaning label names or values cannot be configured).
+  // MTD-custom-labels and MTD-custom-annotations allows to define labels/annotations expected names with config parameters.
+  // MTD-*-(label|annotation) (dynamic rules) allow to define expected values with config parameters.
+  private validateResource(resource: Resource, rule: RuleMetadataWithConfig) {
```

Comment:
I'd maybe consider 4 private functions: `validateRecommendedLabels`, `validateCustomLabels`, `validateCustomAnnotations` and `validateDynamicRule`. Then implement each of these. That might have some slight duplication but will be much easier to understand and make modifications to.

Reply:
I had one, then made two separate methods and then merged again 😓 Agree that it's a bit complicated here, I'll try to split it as you suggested 👍

Reply:
Ok, done in https://github.com/kubeshop/monokle-core/pull/387/commits/6214e04f3cf4c25e2ef259e706e09a7d5d3002b1. I kept error generation logic shared but moved validation part to each separate method so it's now more readable which rule uses which parts of rule config.